File: MarkovChain.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
	"math/rand"
	"time"
	"net/http"
)

func main() {

	t0 := time.Now()
	
	//	bytes := ReadUrl("http://www.gutenberg.org/cache/epub/1661/pg1661.txt")
	bytes, _ := ioutil.ReadFile("PrideAndPrejudice.txt")
	chain := MarkovChain(string(bytes), 100, 2) // ( inputString, chainLength, matchLength )
	
	duration := time.Since(t0)
	fmt.Println(chain)
	fmt.Println("Duration:", duration)
	
}

func MarkovChain( inputString string, chainLen int, matchLen int) []string {

	// MarkovChain( inputString, chainLength, matchLength )
	// Make a markov chain from an input string that's chainLen words long which grows 
	// by matching the last matchLen words in the chain ( generally 2 )

	words := strings.Fields(inputString)
		
	// create chain and fill it with two words from a random location
	rand.Seed(time.Now().Unix())
	chain := []string{}
	pos := rand.Intn(len(words)-matchLen)
	chain = append(chain, words[pos:pos+matchLen]...)			
			
	for i:=1; i<=chainLen; i+=1 {
	
		query := chain [ len(chain)-matchLen: ]
		matches := []string{}	
		
		for i,_ := range words {
			doesItMatch := true
			for j,_ := range query {
				if query[j] != words[i+j] {
					doesItMatch = false
					break
				}
			}
			if doesItMatch {
				matches = append(matches, words[i+2])
			}
		}

		if len(matches) == 0 { 
			break 
		} else if len(matches) == 1 { 
			chain = append(chain, matches[0])
		} else if len(matches) > 1 {
			chain = append( chain, matches[rand.Intn(len(matches)-1)] )
		}
				
	}
	return chain
}

func ReadUrl(url string) []byte {
//	resp, err := http.Get("http://www.gutenberg.org/cache/epub/1661/pg1661.txt")
	resp, err := http.Get(url)
	if err != nil {
		panic(err) 
	}
	defer resp.Body.Close()
	body, err := ioutil.ReadAll(resp.Body)
	return body
}

func MarkovChainStart( arg []byte) []string {

	words := strings.Fields(string(arg))
		
	// create chain and fill it with two words from a random location
	rand.Seed(time.Now().Unix())
	chain := []string{}
	pos := rand.Intn(len(words)-2)
	chain = append(chain, words[pos])
	chain = append(chain, words[pos+1])
			
			
	for i:=1; i<=50; i+=1 {
	
		query := chain [ len(chain)-2: ] // last two words in chain
		matches := []string{}	
		
		for i,_ := range words {
			if words[i] == query[0] && words[i+1] == query[1] {
				matches = append(matches, words[i+2])
			}
		}

		if len(matches) == 0 { 
			break 
		} else if len(matches) == 1 { 
			chain = append(chain, matches[0])
		} else if len(matches) > 1 {
			chain = append( chain, matches[rand.Intn(len(matches)-1)] )
		}
				
	}
	return chain
}

File: MarkovChain.r
#!/usr/local/bin/rebol

comment { http://www.gutenberg.org/ebooks/98.txt.utf-8   51,479 free ebooks }

rebol []

comment {
	Find every word that follows the last two words in the chain 
	and append one randomly to the chain.
	Get two words from a random position in the text to begin.

	markov-words: [
		cache: [] words: terms: chain: pos: matches: match: sources: url: page: body: none
	]
	; unset first words-of markov-words

}

at-random: func [ arg ] [ at arg random length? arg ]

markov-chain: funct [
	"Generate impressively readable gibberish"
	arg
][
	random/seed now
	
	words: parse arg none

	chain: copy/part (skip at-random words -1) 2
	loop 50 [
		pos: words
		query: skip tail chain -2
		matches: collect [
			while [ pos: find/tail pos query ][ keep pos/1 ]
		]
		append chain random/only matches
	]
	return chain
]

demo-markov-chain: does [

	sources: load [
		http://www.rebol.com/docs/words/wwait.html
		http://www.gutenberg.org/cache/epub/1661/pg1661.txt ; "Sherlock Holmes" 
		http://gutenberg.net.au/ebooks01/0100021h.html
	]
	unless value? 'cache [ 
		cache: []
	]
	; "If the cache already exists then don't override it

	url: random/only sources
	page: select cache url
	unless string? page [
		print join "Reading " url
		page: read url
		if find [ %.html %.htm ] suffix? url [
			; "Get stuff between body tags then remove tags"
			body: ""
			parse page [ to "<body" start: thru "<body" copy body thru "body>" ]
			page: rejoin remove-each bit load/markup body [tag? bit] 
		]
		append cache reduce [ url page ]
	]
	print url
	duration: dt [ chain: markov-chain page ]
	print chain
	print [ "Duration:" duration ]
	

]

demo-markov-chain

